File: src/901-1000/995.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <fstream>
#include <sstream>
#include <map>
#include <chrono>
#include <gmpxx.h> // GMP C++ wrapper

using namespace std;

const int MAX_PRIME_SIEVE = 2000000;
const string CACHE_FILE = "min_s_cache.txt";

// Fast modular exponentiation
long long pow_mod(long long base, long long exp, long long mod) {
    long long res = 1;
    base %= mod;
    while (exp > 0) {
        if (exp % 2 == 1) res = (res * base) % mod;
        base = (base * base) % mod;
        exp /= 2;
    }
    return res;
}

// Sieve arrays
vector<bool> is_p_sieve;
vector<int> primes;

void build_sieve() {
    is_p_sieve.assign(MAX_PRIME_SIEVE, true);
    is_p_sieve[0] = is_p_sieve[1] = false;
    for (int i = 2; i * i < MAX_PRIME_SIEVE; i++) {
        if (is_p_sieve[i]) {
            for (int j = i * i; j < MAX_PRIME_SIEVE; j += i) {
                is_p_sieve[j] = false;
            }
        }
    }
    for (int i = 2; i < MAX_PRIME_SIEVE; i++) {
        if (is_p_sieve[i]) primes.push_back(i);
    }
}

// Fallback for primes larger than 2,000,000
bool is_prime_dynamic(long long n) {
    if (n < MAX_PRIME_SIEVE) return is_p_sieve[n];
    if (n % 2 == 0 || n % 3 == 0) return false;
    for (long long i = 5; i * i <= n; i += 6) {
        if (n % i == 0 || n % (i + 2) == 0) return false;
    }
    return true;
}

// Unordered multiplicative factorizations
vector<vector<int>> get_factorizations(int n, int start_factor = 2) {
    vector<vector<int>> res = {{n}};
    for (int i = start_factor; i * i <= n; ++i) {
        if (n % i == 0) {
            for (auto sub : get_factorizations(n / i, i)) {
                vector<int> fact = {i};
                fact.insert(fact.end(), sub.begin(), sub.end());
                res.push_back(fact);
            }
        }
    }
    return res;
}

// Find smallest primitive root generator for p
int primitive_root(int p) {
    if (p == 2) return 1;
    int phi = p - 1;
    vector<int> factors;
    int n = phi;
    for (int i = 2; i * i <= n; ++i) {
        if (n % i == 0) {
            factors.push_back(i);
            while (n % i == 0) n /= i;
        }
    }
    if (n > 1) factors.push_back(n);

    for (int res = 2; res < p; ++res) {
        bool valid = true;
        for (int f : factors) {
            if (pow_mod(res, phi / f, p) == 1) {
                valid = false;
                break;
            }
        }
        if (valid) return res;
    }
    return -1;
}

// Load cache from disk
map<int, string> load_cache() {
    map<int, string> cache;
    ifstream infile(CACHE_FILE);
    string line;
    while (getline(infile, line)) {
        size_t colon_pos = line.find(':');
        if (colon_pos != string::npos) {
            int p = stoi(line.substr(0, colon_pos));
            string s_val = line.substr(colon_pos + 1);
            // strip whitespace
            s_val.erase(remove_if(s_val.begin(), s_val.end(), ::isspace), s_val.end());
            cache[p] = s_val;
        }
    }
    return cache;
}

// Append new result to cache
void save_to_cache(int p, const mpz_class& s) {
    ofstream outfile(CACHE_FILE, ios_base::app);
    outfile << p << ": " << s.get_str() << "\n";
}

// The core algorithm powered by GMP mpz_class
mpz_class find_minimum_s(int p) {
    if (p == 2) return 1;
    
    int g = primitive_root(p);
    vector<vector<int>> factorizations = get_factorizations(p - 1);
    
    // O(1) Pre-bucket
    vector<int> first_prime_for_mod(p, 0);
    int missing = p - 1;
    for (int q : primes) {
        int rem = q % p;
        if (rem != 0 && first_prime_for_mod[rem] == 0) {
            first_prime_for_mod[rem] = q;
            missing--;
            if (missing == 0) break;
        }
    }

    mpz_class min_s = 0;
    bool found_any = false;

    for (const auto& fact : factorizations) {
        vector<int> perm = fact;
        sort(perm.begin(), perm.end()); // Required for next_permutation
        
        do {
            mpz_class s_val = 1;
            int W = 1;
            
            for (int c : perm) {
                int best_q = -1;
                
                // Algebraic jump sizes
                for (int k = 1; k <= (p - 1) / W; ++k) {
                    if (std::gcd(k, c) == 1) {
                        int req_res = pow_mod(g, W * k, p);
                        int q = first_prime_for_mod[req_res];
                        
                        // Dynamic prime fallback if > 2,000,000
                        if (q == 0) {
                            q = req_res;
                            if (q < 2) q += p;
                            while (!is_prime_dynamic(q)) q += p;
                            first_prime_for_mod[req_res] = q;
                        }
                        
                        if (best_q == -1 || q < best_q) best_q = q;
                    }
                }
                
                // s_val *= best_q ^ (c - 1) using GMP
                mpz_class q_pow;
                mpz_ui_pow_ui(q_pow.get_mpz_t(), best_q, c - 1);
                s_val *= q_pow;
                
                W *= c;
            }
            
            if (!found_any || s_val < min_s) {
                min_s = s_val;
                found_any = true;
            }
            
        } while (next_permutation(perm.begin(), perm.end()));
    }
    return min_s;
}

int main() {
    int N = 100;
    
    cout << "Building Prime Sieve..." << endl;
    build_sieve();
    
    cout << "Loading cache..." << endl;
    map<int, string> cache = load_cache();
    
    cout << "Calculating minimum 's' for primes < " << N << "...\n\n";
    
    auto start_time = chrono::high_resolution_clock::now();

    for (int p : primes) {
        if (p >= N) break;
        
        if (cache.count(p)) {
            string s_str = cache[p];
            if (s_str.length() > 30) {
                cout << p << "\t| " << s_str.substr(0, 5) << "..." 
                     << s_str.substr(s_str.length() - 5) 
                     << " (" << s_str.length() << " digits) [Cached]\n";
            } else {
                cout << p << "\t| " << s_str << " [Cached]\n";
            }
        } else {
            mpz_class s = find_minimum_s(p);
            save_to_cache(p, s);
            
            string s_str = s.get_str();
            if (s_str.length() > 30) {
                cout << p << "\t| " << s_str.substr(0, 5) << "..." 
                     << s_str.substr(s_str.length() - 5) 
                     << " (" << s_str.length() << " digits)\n";
            } else {
                cout << p << "\t| " << s_str << "\n";
            }
        }
    }

    auto end_time = chrono::high_resolution_clock::now();
    chrono::duration<double> elapsed = end_time - start_time;
    cout << "\nExecution time: " << elapsed.count() << " seconds\n";

    return 0;
}

File: src/901-1000/995.py
import os
import math
import itertools
import time

CACHE_FILE = "min_s_cache.txt"
MAX_PRIME_SIEVE = 2000000

# 1. Fast Prime Sieve for instant lookups
is_p = [True] * MAX_PRIME_SIEVE
is_p[0] = is_p[1] = False
for i in range(2, int(MAX_PRIME_SIEVE**0.5) + 1):
    if is_p[i]:
        for j in range(i*i, MAX_PRIME_SIEVE, i):
            is_p[j] = False
primes = [i for i, val in enumerate(is_p) if val]

def is_prime_dynamic(n):
    """Fallback primality test if we exceed the 2M sieve."""
    if n < MAX_PRIME_SIEVE: return is_p[n]
    if n % 2 == 0 or n % 3 == 0: return False
    for i in range(5, int(n**0.5) + 1, 6):
        if n % i == 0 or n % (i + 2) == 0: return False
    return True

def get_factorizations(n, start_factor=2):
    """Unordered multiplicative factorizations."""
    res = [[n]]
    for i in range(start_factor, int(n**0.5) + 1):
        if n % i == 0:
            for sub in get_factorizations(n // i, i):
                res.append([i] + sub)
    return res

def primitive_root(p):
    """Finds the smallest primitive root generator for p."""
    if p == 2: return 1
    phi = p - 1
    factors = []
    n = phi
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            factors.append(i)
            while n % i == 0: n //= i
    if n > 1: factors.append(n)

    for res in range(2, p):
        valid = True
        for f in factors:
            if pow(res, phi // f, p) == 1:
                valid = False; break
        if valid: return res
    return None

def load_cache():
    cache = {}
    if os.path.exists(CACHE_FILE):
        with open(CACHE_FILE, "r") as f:
            for line in f:
                if ":" in line:
                    p_str, s_str = line.strip().split(":", 1)
                    cache[int(p_str.strip())] = int(s_str.strip())
    return cache

def save_to_cache(p, s):
    with open(CACHE_FILE, "a") as f:
        f.write(f"{p}: {s}\n")

def find_minimum_s(p):
    if p == 2: return 1
    
    g = primitive_root(p)
    factorizations = get_factorizations(p - 1)
    
    # Pre-bucket the smallest prime for every possible residue modulo p
    # This turns prime searching into an O(1) array lookup
    first_prime_for_mod = [None] * p
    missing = p - 1
    for q in primes:
        rem = q % p
        if rem != 0 and first_prime_for_mod[rem] is None:
            first_prime_for_mod[rem] = q
            missing -= 1
            if missing == 0: break

    min_s = float('inf')

    for fact in factorizations:
        # Test every unique permutation of the subset sizes
        for perm in set(itertools.permutations(fact)):
            s_val = 1
            W = 1
            
            for c in perm:
                best_q = float('inf')
                
                # Algebraic conditions: L = W * k, where gcd(k, c) == 1
                for k in range(1, (p - 1) // W + 1):
                    if math.gcd(k, c) == 1:
                        # Calculate the required residue modulo p
                        req_res = pow(g, W * k, p)
                        
                        # O(1) lookup for the absolute smallest prime with this residue
                        q = first_prime_for_mod[req_res]
                        
                        # Dynamic fallback if prime > 2,000,000 (extremely rare)
                        if q is None:
                            q = req_res if req_res >= 2 else req_res + p
                            while not is_prime_dynamic(q): q += p
                            first_prime_for_mod[req_res] = q
                            
                        if q < best_q:
                            best_q = q
                            
                s_val *= best_q ** (c - 1)
                W *= c
                
            if s_val < min_s:
                min_s = s_val

    return min_s

def main(N):
    print(f"Loading cache and calculating minimum 's' for primes < {N}...\n")
    cache = load_cache()
    
    start_time = time.time()
    
    for p in primes:
        if p >= N: break
        
        if p in cache:
            s = cache[p]
            print(f"{p:<10} | {s} (Cached)")
        else:
            s = find_minimum_s(p)
            save_to_cache(p, s)
            cache[p] = s
            # Format output to gracefully handle massive BigInts
            s_str = str(s)
            if len(s_str) > 30:
                s_str = f"{s_str[:5]}...{s_str[-5:]} ({len(s_str)} digits)"
            print(f"{p:<10} | {s_str}")

    print(f"\nExecution time: {time.time() - start_time:.4f} seconds")

if __name__ == "__main__":
    main(20000)